.github/workflows/dynamic-ci.yml

```diff
@@ -0,0 +1,104 @@
+name: Dynamic CI
+
+on:
+  push:
+    branches:
+      - main
+      - release/[0-9]+.[0-9]+
+    tags:
+      - v[0-9]+.[0-9]+.[0-9]+
+  pull_request:
+    types: [opened, synchronize, reopened, ready_for_review]
+  merge_group:
+    types: [checks_requested]
+
+jobs:
+  filter:
+    name: Filter changed files
+    runs-on: ubuntu-latest
+    permissions:
+      pull-requests: read
+    outputs:
+      docs: ${{ steps.filter.outputs.docs }}
+      code: ${{ steps.filter.outputs.code }}
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v4
+
+      - name: Detect changed files
+        id: filter
+        uses: dorny/paths-filter@v3
+        with:
+          # 'base' and 'ref' are ignored for pull requests
+          # For merge group events, compare against the target branch (main)
+          base: ${{ github.event.merge_group.base_ref }}
+          # For merge group events, use the merge group head ref
+          ref: ${{ github.event.merge_group.head_ref }}
+          filters: |
+            docs:
+              - '.github/workflows/automatic-doc-checks.yml'
+              - 'docs/**'
+            code:
+              - ".github/workflows/linux*"
```

Comment:
Would it make sense to remove the code filter and use !doc in the non-doc workflow dispatch calls? My concern here is that we might forget updating this as time passes, especially on files that have a non-direct influence on the outcome.

Reply:
The disadvantage of doing that is that CI would build Multipass for changes to unrelated files such as `renovate.json5`. We could add another filter for these *meta*-files and call somethings like `!doc && !meta`, but then we would have to remember to maintain that list of files instead of a list of code paths. The `code` directories are all top-level directories so in theory they shouldn't be changing often.

Reply:
After thinking about it again, it wouldn't be hard to notice that the required workflow hasn't run for a PR where it should.